src/sync.rs

```diff
@@ -0,0 +1,302 @@
+use std::collections::{hash_map::Entry, HashMap, VecDeque};
+use std::convert::TryFrom;
+use std::num::NonZeroU32;
+use std::time::Instant;
+
+use rustc_index::vec::{Idx, IndexVec};
+
+use crate::*;
+
+macro_rules! declare_id {
```

Comment:
Why do we need our own macro; `rustc_index` declares `newtype_index!` which is usually used with `IndexVec`?

Reply:
Oh... I guess that's related to the shift by 1... we need to start at 0 for `IndexVec` to work?
First of all, this definitely needs a comment explaining what happens.^^

But secondly... is this really needed? `newtype_index!` have a niche at the end, so we could still use `0xFFFFFFFF` as sentinel value.

Reply:
As a sentinel only `0` works because that is the value with which static initializers initialize memory.